td-layout/src/metadata.rs

```diff
@@ -80,32 +121,125 @@ pub struct TdxMetadataSection {
     pub attributes: u32,
 }
 
+impl TdxMetadataSection {
+    pub fn get_type_name(r#type: u32) -> Option<String> {
+        if r#type >= TDX_METADATA_SECTION_TYPE_MAX {
+            None
+        } else {
+            Some(String::from(
+                TDX_METADATA_SECTION_TYPE_STRS[r#type as usize],
+            ))
+        }
+    }
+
+    pub fn from_bytes(buffer: &[u8; 32]) -> Option<TdxMetadataSection> {
+        let mut metadata_section = TdxMetadataSection::default();
+        metadata_section.data_offset = u32::from_le_bytes(buffer[..4].try_into().unwrap());
+        metadata_section.raw_data_size = u32::from_le_bytes(buffer[4..8].try_into().unwrap());
+        metadata_section.memory_address = u64::from_le_bytes(buffer[8..16].try_into().unwrap());
+        metadata_section.memory_data_size = u64::from_le_bytes(buffer[16..24].try_into().unwrap());
+        metadata_section.r#type = u32::from_le_bytes(buffer[24..28].try_into().unwrap());
+        metadata_section.attributes = u32::from_le_bytes(buffer[28..32].try_into().unwrap());
+
+        if metadata_section.is_valid() {
+            Some(metadata_section)
+        } else {
+            None
+        }
+    }
+
+    pub fn is_valid(&self) -> bool {
+        let mut valid = false;
+        match self.r#type {
+            TDX_METADATA_SECTION_TYPE_BFV => {
+                valid = self.raw_data_size != 0
+                    && self.memory_address != 0
+                    && self.memory_data_size != 0
+                    && self.attributes == TDX_METADATA_ATTRIBUTES_EXTENDMR;
+            }
+
+            TDX_METADATA_SECTION_TYPE_CFV => {
+                valid = self.raw_data_size != 0
+                    && self.memory_address != 0
+                    && self.memory_data_size != 0
+                    && self.attributes == 0;
+            }
+
+            TDX_METADATA_SECTION_TYPE_TD_HOB => {
+                valid = self.data_offset == 0
+                    && self.raw_data_size == 0
+                    && self.memory_address != 0
+                    && self.memory_data_size != 0
+                    && self.attributes == 0;
+            }
+
+            TDX_METADATA_SECTION_TYPE_TEMP_MEM => {
+                valid = self.data_offset == 0
+                    && self.raw_data_size == 0
+                    && self.memory_address != 0
+                    && self.memory_data_size != 0
+                    && self.attributes == 0;
+            }
+
+            TDX_METADATA_SECTION_TYPE_PAYLOAD => {
+                valid = self.data_offset == 0
```

Comment:
I think it is legal to have payload inside of binary. In this case data_offset != 0, raw_data_size != 0